# 1. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [2]:
path = r'/Users/test/Desktop/Data Analysis/11-2024 Instacart Basket Analysis'

In [28]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [29]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 3. Mixed-Type Data

In [30]:
# Create a Dataframe
df_test = pd.DataFrame()

In [31]:
# Create a mixed-type column
df_test['mix'] = ['a', 'b', 1, True]

In [32]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [33]:
# Convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# 4. Missing Values

In [34]:
# Find missing values.(Every missing observation is equal to 1)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [35]:
# Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [36]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [37]:
df_prods.shape

(49693, 5)

In [38]:
# Create a new dataframe without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [39]:
# Check the number of rows in the clean dataframe
df_prods_clean.shape

(49677, 5)

To drop all missing values --> df_prods.dropna(inplace = True)

To drop only the NaNs from a particular column --> df_prods.dropna(subset = [‘product_name’], inplace = True)

# 5. Duplicates

In [40]:
# Create a new subset of 'df_prods_clean' — 'df_dups', containing only rows that are duplicates.
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [41]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [42]:
# Check the current number of rows in 'df_prods_clean' dataframe.
df_prods_clean.shape

(49677, 5)

In [43]:
# Create a new dataframe that doesn’t include the duplicates.
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [44]:
df_prods_clean_no_dups.shape

(49672, 5)

# 6. Exporting Changes

In [45]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Exercise 4.5 Task

## Step 2: Run the df.describe() function on your df_ords dataframe. Interpret the output of this function.

In [46]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

##### 1.'order_id' and 'user_id' should be considered as string values instead of numerical values.
##### 2.'order_number' has a maximum value of 100, which seems very high depending on the time frame.
##### 3.'orders_day_of_week' has a minimum of '0' and a maximum of '6' meaning the seven numbers (0-6) represent days of the week.
##### 4.'order_hour_of_day' has values from 0 to 23 meaning the twenty-four numbers (0-23) represent 24 hours of the day.
##### 5.The minimum value for 'days_since_prior_order' is 0, suggesting multiple orders on the same day.
##### 6.The count of 'days_since_prior_order' is lower compared to the other columns, suggesting there are missing values.

## Step 3: Check for mixed-type data in your df_ords dataframe

In [47]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There's no mixed data present in the df_ords dataframe.

In [48]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## Step 5: Find missing values (takes sum of Trues in a column).

In [49]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The 'days_since_prior_order' column is missing 206209 values.
An explanation for this could be customers' first orders don't have a value.

# Step 6: Address the missing values using an appropriate method.

In [50]:
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [52]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

It seems like all 'order_number' values equal to 1 are missing values in the 'days_since_prior_order' column.

In [53]:
# Filter the dataframe where 'order_number' equals 1 to verify my hypothesis.
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [54]:
# Check if all 'days_since_prior_order' values in these rows are NaN
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [55]:
# Print the result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


In [56]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1.
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [57]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [58]:
# Recheck data.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

## Step 7: Run a check for duplicate values in your df_ords data.

In [59]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [60]:
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

There are no duplicates in this set because it returned empty. 

If there were duplicates I would execute 'df_ords.drop_duplicates()' to create a new dataframe without duplicates.

## Step 9: Export your cleaned df_ords data in your “Prepared Data” folder.

In [61]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

# Revision

In [62]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum value in the 'prices' column 99999 is anormaly high, which can be an error. 

The 'product_id' column has a maximum value of 49688, which doesn't match the count number 49672. This can also be caused by an error or duplicate values.

In [63]:
# Create a complete set of integers from 1 to 49688
df_all_ids = set(range(1, df_prods_clean_no_dups['product_id'].max() + 1))

In [64]:
# Create a set of existing product_id values in the dataframe
df_existing_ids = set(df_prods_clean_no_dups['product_id'].unique())

In [65]:
# Find the missing product IDs
missing_ids = df_all_ids - df_existing_ids

In [67]:
# Print the missing product IDs
print(f"Missing product IDs: {sorted(missing_ids)}")

Missing product IDs: [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 6799, 26519, 38183, 40440]


These results indicate that some product ids are not collected or have been removed already. 

In [68]:
# Now we look at the maximum value of prices.
print(df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999])

       product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0


In [69]:
# The price is obviously an error. We can remove it.
df_prods_no_anormals= df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] != 99999]

In [70]:
df_prods_no_anormals.describe()

product_id      aisle_id  department_id        prices
count  49671.000000  49671.000000   49671.000000  49671.000000
mean   24850.172334     67.762115      11.728856      7.980256
std    14340.795118     38.316100       5.850806     66.952504
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24850.000000     69.000000      13.000000      7.100000
75%    37268.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

There is still anormal value.

In [71]:
# Let's list all the rows with highest prices to look for other anormal values.
df_prods_sorted = df_prods_no_anormals.sort_values(by='prices', ascending=False)

In [72]:
df_prods_sorted.head()

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
9020         9020  Boneless Skinless Chicken Thighs        35             12   
40490       40486                   Chicken Tenders        49             12   
21468       21467            Wild Caught Raw Shrimp        15             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   

        prices  
21554  14900.0  
9020      25.0  
40490     25.0  
21468     25.0  
25580     25.0

It seems that there is only one anormal value left. 

In [73]:
# Remove the row of price 14900.
df_prods_no_anormals_2 = df_prods_no_anormals[df_prods_no_anormals['prices'] != 14900]

In [74]:
df_prods_no_anormals_2.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [75]:
# Export cleaned df_prods to Prepared Data folder.
df_prods_no_anormals_2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_rechecked.csv'))